server/users.ts

```diff
@@ -359,6 +362,9 @@ export class User extends Chat.MessageContext {
 	lastDisconnected: number;
 	lastConnected: number;
 	foodfight?: {generatedTeam: string[], dish: string, ingredients: string[], timestamp: number};
+	// We want this to be null on creation so that the database can be queried only once per user object.
+	friends: Set<string> | null;
+	friendRequests: {sent: Set<string>, received: Set<string> | null} | null;
```

Comment:
I don't think these are good attributes to have on the `User` object.

These are caches, not the source of truth. The database is the source of truth. Ideally, you would query the database every time you wanted to know a friend related fact. In this situation you would never use these attributes.

You have decided that it would be too expensive to query the database every single time. I don't know if that is true or not: SQLite keeps its own cache, the query planning is done beforehand with `prepare`, and there are no IPC costs. But assuming it is, that optimization should not bleed into the interface--nothing that uses the friends list should be able to tell whether you are using this optimization or not.

If you need a cache, use a `Map<User, ...>` local to the chat plugin. And don't have the rest of the code interface with the map directly; when you ask for friends list data, it should check if the data is already present in the cache, and if not, load from the database and store it in the cache.

I suspect this will remove a lot of the `?` problems, and will make it easier to remove the caching/have cache expiration policies if necessary.

Reply:
In the interest of merging in a timely manner, I'd recommend just doing this first without any caching, especially since a limited # of people can use this to start with. Please reach out on Discord if you have any questions.

Reply:
Done. It now checks the database exclusively.